# EJERCICIO - EJEMPLO

## Considera una imagen fotográfica, cárgala usando la librería OpenCV. Pásala a escala de grises usando 1) Y del modelo YUV y 2) el canal I del modelo HSI. Averigua cuál de ellas es la transformación que realiza OpenCV cuando carga una imagen como escala de grises o cuando la transforma con su función cvtColor. ¿Por qué es esa transformación la que se realiza?

Primero cargaremos la imagen a color y la pasamos a escala de grises con los recursos de OpenCV. Iremos sacando por pantalla los distintos datos para familiarizarnos con su estructura.

In [197]:
import numpy as np #Cargamos la librería numpay para trabajar con arrays
import cv2 #Cargamos OpenCV

# Cargamos la imagen tal cual está guardada
img = cv2.imread('airplane.png',-1) #ponemos el nombre de una imagen que se encuentre en el mismo directorio 
#El segundo parámetro de la función cv2.imread() es una etiqueta (flag) que puede ser: 
#  IMREAD_UNCHANGED = -1,
#  IMREAD_GRAYSCALE = 0,
#  IMREAD_COLOR = 1.
img

array([[[ 98,   0, 181],
        [202, 206, 185],
        [197, 207, 165],
        ...,
        [206, 166, 150],
        [202, 164, 150],
        [196, 147, 131]],

       [[108,   0, 141],
        [204, 193, 199],
        [200, 189, 195],
        ...,
        [204, 182, 162],
        [195, 155, 141],
        [197, 139, 119]],

       [[108,   0, 141],
        [203, 196, 197],
        [199, 193, 193],
        ...,
        [209, 195, 174],
        [196, 159, 144],
        [193, 144, 116]],

       ...,

       [[172,   0, 160],
        [212, 215, 210],
        [214, 214, 211],
        ...,
        [191, 185, 181],
        [190, 175, 173],
        [168, 144, 158]],

       [[173,   0, 163],
        [213, 216, 210],
        [215, 215, 210],
        ...,
        [177, 163, 168],
        [190, 182, 184],
        [184, 164, 167]],

       [[ 32, 127,  35],
        [ 34, 123,  33],
        [ 38, 125,  34],
        ...,
        [ 68, 127,  98],
        [ 65, 125, 102],
        [ 54, 117, 107]]

Nótese, que cada fila de la matriz está compuesta por vectores de 3 elementos que son los correspondientes a los canales B, G y R (ya que OpenCV ordena así los canales de color). Ahora pasaremos la imagen a escala de grises. 

In [198]:
#Podemos pasar la imagen a escala de grises usando cvtColor, que sirve en general para transformar la imagen a grises 
#o a otro modelo de color
img_gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
#También la podríamos haber cargado directamente a escala de grises con el "flag" 0 para imread
img_gray

array([[ 65, 199, 193, ..., 166, 164, 148],
       [ 54, 196, 192, ..., 179, 155, 140],
       [ 54, 197, 194, ..., 190, 159, 141],
       ...,
       [ 67, 213, 213, ..., 184, 176, 151],
       [ 68, 214, 214, ..., 166, 184, 167],
       [ 89,  86,  88, ..., 112, 111, 107]], dtype=uint8)

 Ahora cada fila está compuesta por un valor de gris por cada píxel.

In [199]:
# Mostramos las imágenes
cv2.imshow('image',img)
cv2.imshow('image_gray',img_gray)

Ahora separamos los canales y obtenemos, mediante la transformación apropiada, el canal Y del modelo YUV, así como el canal I de HSI.

In [200]:
#Separamos cada matriz, debemos tener en cuenta que la imagen está guardada usando la representación BGR
B, G, R = cv2.split(img) 
#La aplicacion de la fórmula se puede hacer de manera directa gracias a las operaciones sobre arrays de numpy
Y = 0.299 * R + 0.587 * G + 0.114 * B
#pero el resultado serán números decimales por lo que volvemos a pasar al tipo de dato uint8 de las imágenes:
Y8=np.uint8(Y)
cv2.imshow('Y',Y8)
#Del mismo modo calculamos la componente I de la representación HSI
I=0.333*R+0.333*G+0.333*B
#pero el resultado serán números decimales por lo que volvemos a pasar al tipo de dato uint8 de las imágenes:
I8=np.uint8(I)
cv2.imshow('I',I8)

In [201]:
Y # Veo que Y tiene valores decimales 

array([[ 65.291, 199.265, 193.302, ..., 165.776, 164.146, 147.802],
       [ 54.471, 196.048, 192.048, ..., 178.528, 155.374, 139.632],
       [ 54.471, 197.097, 193.684, ..., 190.317, 158.733, 141.214],
       ...,
       [ 67.448, 213.163, 213.103, ..., 184.488, 176.112, 150.922],
       [ 68.459, 213.864, 213.505, ..., 166.091, 183.51 , 167.177],
       [ 88.662,  85.944,  87.873, ..., 111.603, 111.283, 106.828]])

In [202]:
Y8 #vemos cómo Y8 toma el suelo de los valores de Y

array([[ 65, 199, 193, ..., 165, 164, 147],
       [ 54, 196, 192, ..., 178, 155, 139],
       [ 54, 197, 193, ..., 190, 158, 141],
       ...,
       [ 67, 213, 213, ..., 184, 176, 150],
       [ 68, 213, 213, ..., 166, 183, 167],
       [ 88,  85,  87, ..., 111, 111, 106]], dtype=uint8)

Ahora queremos ver si alguna de esas dos imágenes en escala de grises es la que corresponde a la transformación que realiza OpencV internamente cuando usa el flag IMREAD_GRAYSCALE = 0 o bien con cv2.COLOR_BGR2GRAY. Para ello calcularemos el valor absoluto de las diferencias con la imagen img_gray. Ahora bien, como estamos trabajando con arrays de numpy de uint8, al hacer operaciones con ellos, éstas se realizarán en enteros módulo 256. Hay que tener cuidado con eto. 

In [203]:
DI=img_gray-I8
DI

array([[229,   2,   4, ..., 249, 249, 247],
       [228, 254, 254, ..., 253, 248, 245],
       [228, 255,   0, ..., 254, 249, 247],
       ...,
       [213,   1,   1, ..., 255, 253, 251],
       [213,   2,   1, ..., 253, 255, 252],
       [ 25,  23,  23, ...,  15,  14,  15]], dtype=uint8)

Por ejemplo, el valor 229 corresponde a haber obtenido en la diferencia el resultado -27, que es 229 mod 256. Para evitarlo, podemos pasar tanto Y8 como I8 al tipo double y así ya, en el momento que uno de los array de las operaciones es double, las operaciones se harán con números reales. 

In [204]:
Y8=np.double(Y8)
DY=np.abs(Y8-img_gray)

In [205]:
DY

array([[0., 0., 0., ..., 1., 0., 1.],
       [0., 0., 0., ..., 1., 0., 1.],
       [0., 0., 1., ..., 0., 1., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 1.],
       [0., 1., 1., ..., 0., 1., 0.],
       [1., 1., 1., ..., 1., 0., 1.]])

In [192]:
np.max(DY) 

1.0

Enel caso de Y, la mayor diferencia que aparece es de una unidad, que puede deberse al redondeo.

In [206]:
I8=np.double(I8)

In [211]:
DI=np.abs(I8-img_gray)

In [212]:
DI

array([[27.,  2.,  4., ...,  7.,  7.,  9.],
       [28.,  2.,  2., ...,  3.,  8., 11.],
       [28.,  1.,  0., ...,  2.,  7.,  9.],
       ...,
       [43.,  1.,  1., ...,  1.,  3.,  5.],
       [43.,  2.,  1., ...,  3.,  1.,  4.],
       [25., 23., 23., ..., 15., 14., 15.]])

Sin embargo en este caso las diferencias son mayores. Por tanto, la transformación que realiza OpenCV es la del canal Y. Esta transformación a grises es la que se suele usar en todos los paquetes ya que la fórmula realiza un promedio de los tres canales con unos pesos que reflejan la forma de ver del ser humano. Detectamos mucho más el verde, después el rojo y en menor medida el azul.

In [ ]:
# Necesario para cerrar las ventanas:
cv2.waitKey(0)
cv2.destroyAllWindows()

# Importante: interrumpir la ejecución de este bloque para cerrar las ventanas.